Incorporate health and impact damage calculation

ryanc16 opened the pull request.

ryanc16 committed:
Move Accelerometer to Scripts folder

ryanc16 committed:
Update Accelerometer and Monster to try calculating force

ryanc16 committed:
Merge branch 'master' into try-determining-impact-damage

ryanc16 committed:
Create IDamagable interface and Entity class which can be added to game objects which should have health and take damage.

ryanc16 committed:
Make the Monster class extend the Entity class to allow it to have health and take damage

ryanc16 committed:
Update Accelerometer class to continuouslly update the impact vorce along with velocity

ryanc16 committed:
Update the Bird class to have a cooldown between resetting and wait for all entities to stop moving before resetting

ryanc16 committed:
Update the crate prefabs with new values for Entity class.
Add crate_long prefab

ryanc16 committed:
Update BirdRed prefab Accelerometer impactMultiplier value

ryanc16 committed:
Update MonsterPurple prefab with health

ryanc16 committed:
Arrange levels in LevelController prefab

ryanc16 committed:
Update level layouts and add the line renderer for bird trajectory

ryanc16 committed:
Update Editor settings

cjh15hub commented on Assets/Scripts/Accelerometer.cs:
```diff
@@ -1,35 +1,45 @@
+    private void Impact(Collision2D collision)
+    {
+        Accelerometer otherAccel = collision.gameObject.GetComponent<Accelerometer>();
+        _impactForce = otherAccel.impactForce;
```
Check value of otherAccel before using it.

ryanc16 replied:
Done: 698d404cf6a8102f9b7321297d27b2c1ae45435d

cjh15hub commented on Assets/Scripts/Accelerometer.cs:
```diff
@@ -1,35 +1,45 @@
 
 public class Accelerometer : MonoBehaviour
 {
+    private Vector2 _velocity;
```
Please use the public property with a private setter.

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/pull/1/commits/bf5906843a21d2e978aa0e82765b2f23f2a3315e#diff-b9a61da49fb7ea93aad54b4b02c2a551caa0e1f74721482088ba4ee4d1a269e6R7

cjh15hub commented on Assets/Scripts/Accelerometer.cs:
```diff
@@ -1,35 +1,45 @@
     }
-    private Vector2 _velocity;
+
+    private Vector2 _impactForce;
```
Same as above

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/commit/bf5906843a21d2e978aa0e82765b2f23f2a3315e#diff-b9a61da49fb7ea93aad54b4b02c2a551caa0e1f74721482088ba4ee4d1a269e6R12

cjh15hub commented on Assets/Scripts/Accelerometer.cs:
```diff
@@ -1,35 +1,45 @@
-    private Vector2 _velocity;
+
+    private Vector2 _impactForce;
+    public Vector2 impactForce
```
This should be initialized in Awake or Start

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/pull/1/commits/bf5906843a21d2e978aa0e82765b2f23f2a3315e#diff-b9a61da49fb7ea93aad54b4b02c2a551caa0e1f74721482088ba4ee4d1a269e6R22

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/commit/bf5906843a21d2e978aa0e82765b2f23f2a3315e#diff-b9a61da49fb7ea93aad54b4b02c2a551caa0e1f74721482088ba4ee4d1a269e6R22

cjh15hub commented on Assets/Scripts/Bird.cs:
```diff
@@ -114,19 +117,34 @@ private void OnCollisionEnter2D(Collision2D collision)
     {
-        yield return new WaitForSeconds(3);
+        Entity[] allEntities = GameObject.FindObjectsOfType<Entity>();
+        foreach (Entity e in allEntities)
```
You could use a Linq statement if you like

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/pull/1/commits/bf5906843a21d2e978aa0e82765b2f23f2a3315e#diff-c873074d36a3026960706dbe9c6d43acef6e4f0c053e0bc1962b4d9e1bd7716fR130

cjh15hub commented on Assets/Scripts/Bird.cs:
```diff
@@ -114,19 +117,34 @@ private void OnCollisionEnter2D(Collision2D collision)
+        ready = true;
 
-        StartCoroutine(ReadyAfterDelay());
+        //StartCoroutine(ReadyAfterDelay());
```
This was necessary to avoid a bug with dragging him in between life cycle hooks. If the time can be smaller it is okay to decrease the time.

cjh15hub commented on Assets/Scripts/Entity.cs:
```diff
@@ -0,0 +1,82 @@
+{
+    [SerializeField]
+    public float initialHealth = 100;
+    public float health { get { return _health; } set { _health = value; } }
```
`public float health { get ; private set; }`

_health is not needed.

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/pull/1/commits/81f82db7e00559acff3871b77076b9bdb9bcaa43#diff-8446b20d766823004b26f3f1b33f5831cf745c6d4a21828ccea0479ff1c54ce9R8

ryanc16 replied:
Updated: https://github.com/cjh15hub/terribletweeters/pull/1/commits/81f82db7e00559acff3871b77076b9bdb9bcaa43#diff-8446b20d766823004b26f3f1b33f5831cf745c6d4a21828ccea0479ff1c54ce9R8
`_health` was needed to show in the editor as readonly

cjh15hub commented on Assets/Scripts/Monster.cs:
```diff
@@ -1,134 +1,42 @@
-        // if collision is with a bird, always consider it a kill
-        Bird bird = collision.gameObject.GetComponent<Bird>();
-        if (bird)
+        if (isDead)
```
I think this belongs in the base class all together.

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/pull/1/commits/bf5906843a21d2e978aa0e82765b2f23f2a3315e#diff-8446b20d766823004b26f3f1b33f5831cf745c6d4a21828ccea0479ff1c54ce9R46

cjh15hub commented on Assets/Scripts/Entity.cs:
```diff
@@ -0,0 +1,82 @@
+
+    public virtual void DeathRoutine(){}
+
+    private IEnumerator DeactivateAfterDelay()
```
I think this should be called as the default DeathRoutine and DeactivateAfterDelay should accept the parameter for the delay. If a child class overrides the DeathRoutine it is their responsibility to opt into calling DeactiveAfterDealy(time) now.

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/pull/1/commits/81f82db7e00559acff3871b77076b9bdb9bcaa43#diff-8446b20d766823004b26f3f1b33f5831cf745c6d4a21828ccea0479ff1c54ce9R92

cjh15hub commented on Assets/Scripts/Monster.cs:
```diff
@@ -1,134 +1,42 @@
         spriteRenderer.sprite = deadSprite;
         poofParticleSystem.Play();
-
-        StartCoroutine(DeactivateAfterDelay());
```
Based on my comment above with the DeactivateAfterDelay change, can we pass in the "animation" time from the poofParticleSystem instead of a hard coded 1.5f?

ryanc16 replied:
Done: https://github.com/cjh15hub/terribletweeters/commit/81f82db7e00559acff3871b77076b9bdb9bcaa43#diff-021dc1e9eb0584e8b459501c5c0ad3d7dfab210bd7cd63a72fb3ee576efbb985R26

ryanc16 committed:
Cleanup getters and setters.
Cleanup inheritance between Entity and Monster.
Use Linq for EntitiesAreMoving.

ryanc16 committed:
Update OnDeath routine for entity to use a dynamic delay.
Update Monster script to use particle system duration for deactivation on death delay.

ryanc16 committed:
add check for otherAccel
